Add support for django.utils.functional.cached_property

christopherjwang opened the pull request:
Howdy, I think this is what is needed for #9. It passes `pytest`, `flake`, and `mypy`. I'm not sure if I completely mangled the `Pipfile.lock` file though.

<img width="473" alt="screen shot 2018-01-03 at 10 51 05 pm" src="https://user-images.githubusercontent.com/4838266/34550705-a1364742-f0d8-11e7-8371-6af53038ae49.png">

carljm commented on monkeytype/util.py:
```diff
@@ -35,17 +40,19 @@ def get_func_in_module(module: str, qualname: str) -> Callable:
     func = inspect.unwrap(func)  # type: ignore
     if isinstance(func, types.MethodType):
-        func = func.__func__
+        return func.__func__
```
Why this change (and below)?

christopherjwang replied:
Personally, I think it's a lot easier to trace. Once we pass certain conditions there's really no point in continuing down into the rest of the function if we can just exit out early with a known result. I think it could also prevent future people from altering `func` later if we explicitly declare the exact object we want to return with an early return statement.

I'm perfectly okay reverting this change if you disagree.

carljm replied:
Let's revert, it's not related to the changes in this diff. It's a judgment call whether later additions to this function are more likely to be things that should or shouldn't also apply to these unwrapped values, I can envision either possibility. If they shouldn't apply, the existing `elif` chain pattern handles that fine. I prefer to err on the side of single return per function, early returns are easy to miss.

carljm commented:
The reason for the "changes requested" was for tests that exercise the changes to `get_func_in_module` and  `get_func_in_mro`, in addition to the test you added to exercise the change in `stubs.py`.

Thanks again for the pull request!

christopherjwang commented:
Sure thing! I added the tests and reverted the early returns.

mpage commented:
@carljm - +1 for pluggable property/descriptor support. That's going to require some refactoring and thought, so I'm on board with merging this as-is.

@christopherjwang - Can you add a test to verify the generated stub for a function that is decorated with `cached_property`? Based on the screenshot, it looks like the `self` argument is annotated with a type, which it shouldn't be.

carljm commented:
@mpage filed #47 for pluggable property/descriptor support.

Good catch on the self annotation! @christopherjwang can you take a look at why that's happening?

Already-added tests look good.

christopherjwang commented:
The issue was not including `FunctionKind.DJANGO_CACHED_PROPERTY` in `_KIND_WITH_SELF` [here](https://github.com/Instagram/MonkeyType/pull/46/files#diff-17f06850656e0e2dbfd8fd493eb7402dL74). I've fixed it and added more tests, tests should be everywhere the `@property` decorator is tested.

<img width="385" alt="screen shot 2018-01-06 at 3 53 02 pm" src="https://user-images.githubusercontent.com/4838266/34645092-b3c4c744-f2f9-11e7-82f4-ab0cc90a1683.png">
